File: determinant.cpp
#include <iostream>
#include <iomanip>
#include <cmath>

using namespace std;

#define MAX 20

void PrintMatrix(double[][MAX], int);
bool IsAllZero(double[][MAX], int);
int ArgMax(double[][MAX], int, int, int);
double Diagonal(double[][MAX], int);
void SwapRows(double[][MAX], int, int, int);
int GaussElimination(double[][MAX], int);
double Determinant(double[][MAX], int);

int main() {
	ios::sync_with_stdio(false);
	int n;
	double matrix[20][20];
	cout << "Set matrix dimension\n";
	do cin >> n; while (n < 2 || n > 20);
	cout << "Input matrix numbers\n";
	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			cin >> matrix[i][j];
	PrintMatrix(matrix, n);
	cout << setprecision(5) << fixed << "determinant = " << Determinant(matrix, n) << endl;
	cin.get();
}

void PrintMatrix(double matrix[][MAX], int n) {
	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			cout << matrix[i][j] << "\t\n"[n-j==1];
}

bool IsAllZero(double matrix[][MAX], int n) {
	for (int i = 0; i < n; i++)
		if (matrix[i][0] != 0) return false;
	return true;
}

int ArgMax(double matrix[][MAX], int h, int k, int n) {
	int max = fabs(matrix[h][k]);
	int index = h;
	for (int i = h + 1; i < n; i++) {
		if (fabs(matrix[i][k]) > max) {
			max = fabs(matrix[i][k]);
			index = i;
		}
	}
	return index;
}

double Diagonal(double matrix[][MAX], int n) {
	double det = 1;
	for (int i = 0; i < n; i++)
		det *= matrix[i][i];
	return det;
}

void SwapRows(double mat[][MAX], int i, int j, int n) {
	for (int k = 0; k < n; k++) {
		double temp = mat[i][k];
		mat[i][k] = mat[j][k];
		mat[j][k] = temp;
	}
}

int GaussElimination(double matrix[][MAX], int n) {
	int swapCount = 0;
	for (int h = 0, k = 0; h < n && k < n; k++) {
		int i_max = ArgMax(matrix, h, k, n);
		if (matrix[i_max][k] == 0) continue;
		else {
			if (i_max != h) {
				SwapRows(matrix, h, i_max, n);
				swapCount++;
			}
			for (int i = h + 1; i < n; i++) {
				double f = matrix[i][k] / matrix[h][k];
				matrix[i][k] = 0;
				for (int j = k + 1; j < n; j++)
					matrix[i][j] -= matrix[h][j] * f;
			}
			h++;
		}
	}
	return swapCount;
}

double Determinant(double matrix[][MAX], int n) {
	if (IsAllZero(matrix, n)) return 0;
	int swapCount = GaussElimination(matrix, n);
	double result = Diagonal(matrix, n) * pow(-1, swapCount) + 0;
	return result;
}

File: determinant.js
const Matrix = (() => {
    "use strict";
    
    const _isArray = matrix => Array.isArray(matrix);
        
    const _isMatrix = matrix => 
        _isArray(matrix) && matrix.every(array => _isArray(array));
        
    const _isSquare = matrix =>
        matrix.every(array => array.length === matrix.length);
    
    const _isValidMatrix = matrix => {
        if (!_isMatrix(matrix)) throw new Error('Must be a matrix');
        if (!_isSquare(matrix)) throw new Error('Must be a square matrix');
        return true;
    };
    
    const _fix = number => +number.toFixed(10);
    
    const _isAllZero = matrix =>
        matrix.every(row => !row[0]);
    
    const _argmax = (matrix, h,  k) => {
        const length = matrix.length;
        let max = Math.abs(matrix[h][k]);
        let index = h;
        for (let i = h + 1; i < length; i++)
            if (Math.abs(matrix[i][k]) > max) {
                max = Math.abs(matrix[i][k]);
                index = i;
            }
        return index;
    };
    
    const _diagonal = matrix => {
        let det = 1;
        for (let i = 0; i < matrix.length; i++)
            det *= matrix[i][i];
        return det;
    };
    
    const _swapRows = (matrix, i, j) =>
        [matrix[i], matrix[j]] = [matrix[j], matrix[i]];
    
    const _gaussElimination = (matrix, counter) => {
        const length = matrix.length;
        for (let h = 0, k = 0; h < length && k < length; k++) {
            let i_max = _argmax(matrix, h, k);
            if (matrix[i_max][k] === 0) continue;
            else {
                if (i_max !== h) {
                    _swapRows(matrix, h, i_max);
                    counter.swapped++;
                }
                for (let i = h + 1; i < length; i++) {
                    let f = matrix[i][k] / matrix[h][k];
                    matrix[i][k] = 0;
                    for (let j = k + 1; j < length; j++)
                        matrix[i][j] -= matrix[h][j] * f;
                }
                h++;
            }
        }
        return matrix;
    };
    
    const countDeterminant = matrix => {
        if (!_isValidMatrix(matrix)) return null;
        if (_isAllZero(matrix)) return 0;
        const counter = { swapped: 0 };
        matrix = _gaussElimination(matrix, counter);
        let result = _diagonal(matrix) * (-1) ** counter.swapped + 0;
        return _fix(result);
    };
    
    return {
        det: countDeterminant
    };

})();

export default Matrix;
